File: ios/App/App/IsolatedModules/IsolatedModules.swift
//
//  IsolatedModules.swift
//  App
//
//  Created by Julia Samol on 08.09.22.
//

import Foundation
import Capacitor
import WebKit

@objc(IsolatedModules)
public class IsolatedModules: CAPPlugin {
    private let fileExplorer: FileExplorer = .shared
    private lazy var jsEvaluator: JSEvaluator = .init(fileExplorer: fileExplorer)
    
    @objc func previewModule(_ call: CAPPluginCall) {
        call.assertReceived(forMethod: "previewModule", requiredParams: Param.PATH, Param.DIRECTORY)
        
        do {
            guard let path = call.path, let directory = call.directory else {
                throw Error.invalidData
            }
            
            Task {
                do {
                    let module = try fileExplorer.loadPreviewModule(atPath: path, locatedIn: directory)
                    let manifest = try fileExplorer.readModuleManifest(module)
                    let moduleJSON = try await jsEvaluator.evaluatePreviewModule(module)
                    
                    call.resolve([
                        "module": moduleJSON,
                        "manifest": try JSONSerialization.jsonObject(with: manifest)
                    ])
                } catch {
                    call.reject("Error: \(error)")
                }
            }
        } catch {
            call.reject("Error: \(error)")
        }
    }
    
    @objc func registerModule(_ call: CAPPluginCall) {
        call.assertReceived(forMethod: "registerModule", requiredParams: Param.IDENTIFIER, Param.PROTOCOL_IDENTIFIERS)
        
        do {
            guard let identifier = call.identifier, let protocolIdentifiers = call.protocolIdentifiers else {
                throw Error.invalidData
            }
            
            Task {
                do {
                    let module = try fileExplorer.loadInstalledModule(identifier)
                    await jsEvaluator.registerModule(module, forProtocols: protocolIdentifiers)
                    
                    call.resolve()
                } catch {
                    call.reject("Error: \(error)")
                }
            }
        } catch {
            call.reject("Error: \(error)")
        }
    }
    
    @objc func loadModules(_ call: CAPPluginCall) {
        Task {
            do {
                let protocolType = call.protocolType
                let modules: [JSModule] = try fileExplorer.loadAssetModules() + (try fileExplorer.loadInstalledModules())
                
                call.resolve(try await jsEvaluator.evaluateLoadModules(modules, for: protocolType))
            } catch {
                call.reject("Error: \(error)")
            }
        }
    }
    
    @objc func callMethod(_ call: CAPPluginCall) {
        call.assertReceived(forMethod: "callMethod", requiredParams: Param.TARGET, Param.METHOD)
        
        do {
            guard let target = call.target, let method = call.method else {
                throw Error.invalidData
            }
            
            Task {
                do {
                    switch target {
                    case .offlineProtocol:
                        call.assertReceived(forMethod: "callMethod", requiredParams: Param.PROTOCOL_IDENTIFIER)
                        
                        guard let protocolIdentifier = call.protocolIdentifier else {
                            throw Error.invalidData
                        }
                        
                        let args = call.args
                        
                        call.resolve(
                            try await jsEvaluator.evaluateCallOfflineProtocolMethod(method, ofProtocol: protocolIdentifier, withArgs: args)
                        )
                    case .onlineProtocol:
                        call.assertReceived(forMethod: "callMethod", requiredParams: Param.PROTOCOL_IDENTIFIER)
                        
                        guard let protocolIdentifier = call.protocolIdentifier else {
                            throw Error.invalidData
                        }
                        
                        let args = call.args
                        let networkID = call.networkID
                        
                        call.resolve(
                            try await jsEvaluator.evaluateCallOnlineProtocolMethod(method, ofProtocol: protocolIdentifier, onNetwork: networkID, withArgs: args)
                        )
                    case .blockExplorer:
                        call.assertReceived(forMethod: "callMethod", requiredParams: Param.PROTOCOL_IDENTIFIER)
                        
                        guard let protocolIdentifier = call.protocolIdentifier else {
                            throw Error.invalidData
                        }
                        
                        let args = call.args
                        let networkID = call.networkID
                        
                        call.resolve(
                            try await jsEvaluator.evaluateCallBlockExplorerMethod(method, ofProtocol: protocolIdentifier, onNetwork: networkID, withArgs: args)
                        )
                    case .v3SerializerCompanion:
                        call.assertReceived(forMethod: "callMethod", requiredParams: Param.MODULE_IDENTIFIER)
                        
                        guard let moduleIdentifier = call.moduleIdentifier else {
                            throw Error.invalidData
                        }
                        
                        let args = call.args
                        
                        call.resolve(
                            try await jsEvaluator.evaluateCallV3SerializerCompanionMethod(method, ofModule: moduleIdentifier, withArgs: args)
                        )
                    }
                } catch {
                    call.reject("Error: \(error)")
                }
            }
        } catch {
            call.reject("Error: \(error)")
        }
    }
    
    struct Param {
        static let PATH = "path"
        static let DIRECTORY = "directory"
        static let IDENTIFIER = "identifier"
        static let PROTOCOL_IDENTIFIERS = "protocolIdentifiers"
        static let PROTOCOL_TYPE = "protocolType"
        static let TARGET = "target"
        static let METHOD = "method"
        static let ARGS = "args"
        static let PROTOCOL_IDENTIFIER = "protocolIdentifier"
        static let MODULE_IDENTIFIER = "moduleIdentifier"
        static let NETWORK_ID = "networkId"
    }
    
    enum Error: Swift.Error {
        case invalidData
    }
}

private extension CAPPluginCall {
    var path: String? { return getString(IsolatedModules.Param.PATH) }
    
    var directory: Directory? {
        guard let directory = getString(IsolatedModules.Param.DIRECTORY) else { return nil }
        return .init(rawValue: directory)
    }
    
    var identifier: String? { return getString(IsolatedModules.Param.IDENTIFIER) }
    var protocolIdentifiers: [String]? {
        return getArray(IsolatedModules.Param.PROTOCOL_IDENTIFIERS)?.compactMap {
            if let string = $0 as? String {
                return string
            } else {
                return nil
            }
        }
    }
    
    var protocolType: JSProtocolType? {
        guard let protocolType = getString(IsolatedModules.Param.PROTOCOL_TYPE) else { return nil }
        return .init(rawValue: protocolType)
    }
    
    var target: JSCallMethodTarget? {
        guard let target = getString(IsolatedModules.Param.TARGET) else { return nil }
        return .init(rawValue: target)
    }
    
    var method: String? { return getString(IsolatedModules.Param.METHOD) }
    var args: JSArray? { return getArray(IsolatedModules.Param.ARGS)}
    
    var protocolIdentifier: String? { return getString(IsolatedModules.Param.PROTOCOL_IDENTIFIER) }
    var moduleIdentifier: String? { return getString(IsolatedModules.Param.MODULE_IDENTIFIER) }
    
    var networkID: String? { return getString(IsolatedModules.Param.NETWORK_ID) }
}

File: android/app/src/main/java/it/airgap/vault/plugin/isolatedmodules/IsolatedModules.kt
package it.airgap.vault.plugin.isolatedmodules

import androidx.lifecycle.lifecycleScope
import com.getcapacitor.JSArray
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import it.airgap.vault.plugin.isolatedmodules.js.*
import it.airgap.vault.util.*
import kotlinx.coroutines.Deferred
import kotlinx.coroutines.launch

@CapacitorPlugin
class IsolatedModules : Plugin() {
    private val jsEvaluator: Deferred<JSEvaluator> = ExecutableDeferred { JSEvaluator(context, fileExplorer) }
    private val fileExplorer: FileExplorer by lazy { FileExplorer(context) }

    @PluginMethod
    fun previewDynamicModule(call: PluginCall) {
        call.executeCatching {
            assertReceived(Param.PATH, Param.DIRECTORY)

            activity.lifecycleScope.launch {
                executeCatching {
                    val module = fileExplorer.loadPreviewModule(path, directory)
                    val manifest = fileExplorer.readModuleManifest(module)
                    val moduleJson = jsEvaluator.await().evaluatePreviewModule(module)

                    resolve(
                        JSObject(
                            """
                        {
                            "module": $moduleJson,
                            "manifest": ${JSObject(manifest.decodeToString())}
                        }
                    """.trimIndent()
                        )
                    )
                }
            }
        }
    }

    @PluginMethod
    fun registerDynamicModule(call: PluginCall) {
        call.executeCatching {
            assertReceived(Param.IDENTIFIER, Param.PROTOCOL_IDENTIFIERS)

            activity.lifecycleScope.launch {
                executeCatching {
                    val module = fileExplorer.loadInstalledModule(identifier)
                    jsEvaluator.await().registerModule(module, protocolIdentifiers)

                    resolve()
                }
            }
        }
    }

    @PluginMethod
    fun removeDynamicModules(call: PluginCall) {
        activity.lifecycleScope.launch {
            call.executeCatching {
                val jsEvaluator = jsEvaluator.await()

                identifiers?.let {
                    fileExplorer.removeInstalledModules(it)
                    jsEvaluator.deregisterModules(it)
                } ?: run {
                    fileExplorer.removeAllInstalledModules()
                    jsEvaluator.deregisterAllModules()
                }

                resolve()
            }
        }
    }

    @PluginMethod
    fun loadAllModules(call: PluginCall) {
        activity.lifecycleScope.launch {
            call.executeCatching {
                val modules = fileExplorer.loadAssetModules() + fileExplorer.loadInstalledModules()

                resolve(jsEvaluator.await().evaluateLoadModules(modules, protocolType))
            }
        }
    }

    @PluginMethod
    fun callMethod(call: PluginCall) {
        call.executeCatching {
            assertReceived(Param.TARGET, Param.METHOD)

            activity.lifecycleScope.launch {
                executeCatching {
                    val value = when (target) {
                        JSCallMethodTarget.OfflineProtocol -> {
                            assertReceived(Param.PROTOCOL_IDENTIFIER)
                            jsEvaluator.await().evaluateCallOfflineProtocolMethod(method, args, protocolIdentifier)
                        }
                        JSCallMethodTarget.OnlineProtocol -> {
                            assertReceived(Param.PROTOCOL_IDENTIFIER)
                            jsEvaluator.await().evaluateCallOnlineProtocolMethod(method, args, protocolIdentifier, networkId)
                        }
                        JSCallMethodTarget.BlockExplorer -> {
                            assertReceived(Param.PROTOCOL_IDENTIFIER)
                            jsEvaluator.await().evaluateCallBlockExplorerMethod(method, args, protocolIdentifier, networkId)
                        }
                        JSCallMethodTarget.V3SerializerCompanion -> {
                            assertReceived(Param.MODULE_IDENTIFIER)
                            jsEvaluator.await().evaluateCallV3SerializerCompanionMethod(method, args, moduleIdentifier)
                        }
                    }
                    resolve(value)
                }
            }
        }
    }

    override fun handleOnDestroy() {
        super.handleOnDestroy()
        activity.lifecycleScope.launch {
            jsEvaluator.await().destroy()
        }
    }

    private val PluginCall.path: String
        get() = getString(Param.PATH)!!

    private val PluginCall.directory: Directory
        get() = getString(Param.DIRECTORY)?.let { Directory.fromString(it) }!!

    private val PluginCall.identifier: String
        get() = getString(Param.IDENTIFIER)!!

    private val PluginCall.identifiers: List<String>?
        get() = getArray(Param.PROTOCOL_IDENTIFIERS, null)?.toList()

    private  val PluginCall.protocolIdentifiers: List<String>
        get() = getArray(Param.PROTOCOL_IDENTIFIERS).toList()

    private val PluginCall.protocolType: JSProtocolType?
        get() = getString(Param.PROTOCOL_TYPE)?.let { JSProtocolType.fromString(it) }

    private val PluginCall.target: JSCallMethodTarget
        get() = getString(Param.TARGET)?.let { JSCallMethodTarget.fromString(it) }!!

    private val PluginCall.method: String
        get() = getString(Param.METHOD)!!

    private val PluginCall.args: JSArray?
        get() = getArray(Param.ARGS, null)

    private val PluginCall.protocolIdentifier: String
        get() = getString(Param.PROTOCOL_IDENTIFIER)!!

    private val PluginCall.moduleIdentifier: String
        get() = getString(Param.MODULE_IDENTIFIER)!!

    private val PluginCall.networkId: String?
        get() = getString(Param.NETWORK_ID)

    private object Param {
        const val PATH = "path"
        const val DIRECTORY = "directory"
        const val IDENTIFIER = "identifier"
        const val IDENTIFIERS = "identifiers"
        const val PROTOCOL_IDENTIFIERS = "protocolIdentifiers"
        const val PROTOCOL_TYPE = "protocolType"
        const val TARGET = "target"
        const val METHOD = "method"
        const val ARGS = "args"
        const val PROTOCOL_IDENTIFIER = "protocolIdentifier"
        const val MODULE_IDENTIFIER = "moduleIdentifier"
        const val NETWORK_ID = "networkId"
    }
}